rfd/0199-teleport-key-agent.md

```diff
@@ -0,0 +1,514 @@
+
+By running all login state storage through the agent, dependent clients can
+sync their login state without racing over file storage.
+
+This mode would be very useful for Teleport Connect to avoid state drifts
+reflected in the UI. In fact, the Teleport Connect UI would act as a dependent
+client, listening through `tshd` for any login state changes to respond to.
+
+Note: this mode is a stretch goal and is not scheduled to be completed, but
+this RFD would be incomplete if it did not at least consider this future work.
+
+### Client interface
+
+In order to utilize the Teleport Key Agent, dependent clients will need to
+interface with the agent as opposed to file storage or in-memory storage. Agent
+storage will be added as a new type of storage, with a `Keystore` implementation
+for [signing mode](#signing-mode) and a `ProfileStore` and `TrustedCertsStore`
+implementation for [login sync mode](#login-sync-mode-future-work).
+
+### Security
+
+#### Local key agent
+
+For the intended use case of using Teleport Key Agent as a local key agent,
+there is not much of concern to consider:
+
+* The Teleport Key Agent acts as a proxy for normal file storage, and as such
+is protected through file permissions on the unix socket it's served on:
+`$TELEPORT_HOME/agent.sock`.
+* Even in login sync mode, the agent never exposes the underlying private key
+material, only an interface to sign with.
+
+#### Key agent forwarding
+
+We must also consider the unintended use case of forwarding the agent over ssh
+with unix domain socket forwarding. Like ssh agent forwarding, this is an
+insecure use case which will be strongly advised against. However, it cannot
+be entirely avoided, the same way that we can not stop a user from running
+`tsh scp $HOME/.tsh server01:`.
+
+As usual remote attacks are best mitigated through phish-proof authentication.
+If the following features are enabled, the impact of a compromised forwarded
+Teleport Key Agent will be largely mitigated:
+
+* Per-session MFA or Hardware Key Support (pin/touch)
+* MFA for Admin Actions
+
+If deemed necessary, we could put the agent behind an extra
+[confirmation layer](#confirmation-prompt-future-work). This confirmation layer
+would require dependent clients to perform a confirmation action (y/N, pin, or
+touch) to allow the dependent client to utilize the agent.
+
+### Configuration
+
+See [running the agent](#running-the-agent) for how to configure Teleport
+Connect to run the Teleport Key Agent.
+
+### Proto specification
+
+```proto
+// KeyAgentService provides a Teleport key agent service, allowing multiple Teleport client
+// processes to share a single instance of a Teleport key. This is useful for PIV keys which
+// can only be accessed by one process at a time.
+service KeyAgentService {
+  // Query checks whether the provided key index has an existing keyring and
+  // returns the public keys.
+  rpc Query(QueryRequest) returns (QueryResponse) {}
```

Comment:
nit: Use a less generic RPC name? What is it that we are querying?

I would suggest an entity-based design: "rpc GetKey(GetKeyRequest) returns (Key)" instead of "Query".